ide/che-core-ide-app/src/main/java/org/eclipse/che/ide/part/explorer/project/ProjectExplorerViewImpl.java

```diff
@@ -555,16 +514,116 @@ public boolean isShowHiddenFiles() {
+        }
+
+        final ProjectDescriptor updatedDescriptor = event.getUpdatedProjectDescriptor();
+
+        DtoUtil.fill(updatedDescriptor, node.getData());
+
+        if (!tree.getNodeStorage().reIndexNode(oldNodeId, node)) {
+            Log.info(getClass(), "Node wasn't re-indexed");
+        }
+
+        for (Map.Entry<String, Node> entry : oldIdToNode.entrySet()) {
+            if (!tree.getNodeStorage().reIndexNode(entry.getKey(), entry.getValue())) {
+                Log.info(getClass(), "Node wasn't re-indexed");
+            }
+        }
+
+        tree.refresh(node);
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public void onResourceRenamedEvent(ResourceNodeRenamedEvent event) {
+
+        //Here we have old node with old data object and new renamed data object
+        //so we need fetch storable path from it and call tree to find node by this
+        //storable path. When tree will find our node by path, the old node will be
+        //deleted automatically. Otherwise if we can't determine storable path from
+        //node, then we just take parent node and try navigate on it in the tree.
+
+        final String oldNodeId = tree.getNodeStorage().getKeyProvider().getKey(event.getNode());
+
+        Map<String, Node> oldIdToNode = new HashMap<>();
+        for (Node n : tree.getNodeStorage().getAllChildren(event.getNode())) {
+            oldIdToNode.put(tree.getNodeStorage().getKeyProvider().getKey(n), n);
+        }
+
+        if (event.getNode() instanceof ItemReferenceBasedNode) {
+            ItemReference newDTO = (ItemReference)event.getNewDataObject();
+            ItemReferenceBasedNode node = (ItemReferenceBasedNode)event.getNode();
+
+            DtoUtil.fill(newDTO, node.getData());
+        } else if (event.getNode() instanceof ModuleDescriptorNode) {
+            ProjectDescriptor newDTO = (ProjectDescriptor)event.getNewDataObject();
+            ModuleDescriptorNode node = (ModuleDescriptorNode)event.getNode();
+
+            DtoUtil.fill(newDTO, node.getData());
+        }
+
+        if (!tree.getNodeStorage().reIndexNode(oldNodeId, event.getNode())) {
+            Log.info(getClass(), "Node wasn't re-indexed");
+        }
+
+        for (Map.Entry<String, Node> entry : oldIdToNode.entrySet()) {
+            if (!tree.getNodeStorage().reIndexNode(entry.getKey(), entry.getValue())) {
+                Log.info(getClass(), "Node wasn't re-indexed");
+            }
+        }
+
+        tree.refresh(event.getNode());
+
+        if (!event.getNode().isLeaf()) {
+            tree.getNodeLoader().loadChildren(event.getNode(), true);
+        }
+
+        //here we have possible odd behaviour: after renaming directory we should perform checking structure of the expanded directories
+        //cause some of them after renaming may become source directory and we need to replace them with correct nodes, possible solution
+        //is to use node interceptors to intercept expanded nodes
+    }
+
```

Comment:
It seems that this method must be located on presenter. The view must be responsible only for view representation.
tree.getNodeStorage(),event.getNode() We can get this object once and use reference to it. I think this will be more readable.